os/net/nbr-table.c

```diff
@@ -179,101 +168,148 @@ nbr_set_bit(uint8_t *bitmap, nbr_table_t *table, nbr_table_item_t *item, int val
 }
 /*---------------------------------------------------------------------------*/
 static void
-remove_key(nbr_table_key_t *least_used_key)
+remove_key(nbr_table_key_t *key, bool do_free)
 {
   int i;
   for(i = 0; i < MAX_NUM_TABLES; i++) {
     if(all_tables[i] != NULL && all_tables[i]->callback != NULL) {
       /* Call table callback for each table that uses this item */
-      nbr_table_item_t *removed_item = item_from_key(all_tables[i], least_used_key);
+      nbr_table_item_t *removed_item = item_from_key(all_tables[i], key);
       if(nbr_get_bit(used_map, all_tables[i], removed_item) == 1) {
         all_tables[i]->callback(removed_item);
       }
     }
   }
-  /* Empty used map */
-  used_map[index_from_key(least_used_key)] = 0;
+  /* Empty used and locked map */
+  used_map[index_from_key(key)] = 0;
+  locked_map[index_from_key(key)] = 0;
   /* Remove neighbor from list */
-  list_remove(nbr_table_keys, least_used_key);
+  list_remove(nbr_table_keys, key);
+  if(do_free) {
+    /* Release the memory */
+    memb_free(&neighbor_addr_mem, key);
+  }
 }
 /*---------------------------------------------------------------------------*/
-static nbr_table_key_t *
-nbr_table_allocate(nbr_table_reason_t reason, void *data)
+int
+nbr_table_count_entries(void)
 {
-  nbr_table_key_t *key;
-  int least_used_count = 0;
-  nbr_table_key_t *least_used_key = NULL;
-
-  key = memb_alloc(&neighbor_addr_mem);
-  if(key != NULL) {
-    return key;
-  } else {
-#ifdef NBR_TABLE_FIND_REMOVABLE
-    const linkaddr_t *lladdr;
-    lladdr = NBR_TABLE_FIND_REMOVABLE(reason, data);
-    if(lladdr == NULL) {
-      /* Nothing found that can be deleted - return NULL to indicate failure */
-      PRINTF("*** Not removing entry to allocate new\n");
-      return NULL;
-    } else {
-      /* used least_used_key to indicate what is the least useful entry */
-      int index;
-      int locked = 0;
-      if((index = index_from_lladdr(lladdr)) != -1) {
-        least_used_key = key_from_index(index);
-        locked = locked_map[index];
-      }
-      /* Allow delete of locked item? */
-      if(least_used_key != NULL && locked) {
-        PRINTF("Deleting locked item!\n");
-        locked_map[index] = 0;
-      }
+  int i;
+  int count = 0;
+  for(i = 0; i < NBR_TABLE_MAX_NEIGHBORS; i++) {
+    if(used_map[i] > 0) {
+      count++;
     }
-#endif /* NBR_TABLE_FIND_REMOVABLE */
+  }
+  return count;
+}
+/*---------------------------------------------------------------------------*/
+static int
+used_count(const linkaddr_t *lladdr)
+{
+  int item_index = index_from_lladdr(lladdr);
+  int used = used_map[item_index];
```

Comment:
Should there not be a check if lladdr is not found (`index_from_lladdr()` returns `-1`)?

Reply:
I think it should not happen as this is only called on existing entries by `nbr_table_gc_get_worst()`, but I added the check anyway, for good measure.

Reply:
Yes, I agree but `nbr_table_gc_get_worst()` is a non-static function that potentially could be called by an application. Thanks for fixing.